### DISCUSSION TITLE
BPA Perspectives

### DISCUSSION THREAD

**Discussion by nexus150** [TabularEditor/TabularEditor3#514] [Ideas]
Would it be feasible to have a kind of BPA perspectives in both TE2 and TE3? the idea would be that I can apply only a subset of rules to a model, without having to ignore them in each model, for example, in Power BI pro models rules like "Remove Data sources not referenced by any partitions" or "Partition names should match table name for single partition tables" do not make sense, and ignoring them implies a repetitive work in each model and besides filling them with annotations.

I ask this because I have some versions of the BPA_RULES specific for local developments in Power BI Desktop for basic users, and I asked Michael Kovalsky of Microsoft if we could publish it in the official repo of rules, and what he tells me is that maintaining several files that share much of the same rules, may not be ideal, and may be susceptible to errors in addition to the rework involved.

would be simply to be able to store a variant of the rules that apply or not, currently you can activate / deactivate the rules individually and the idea would be to be able to store a specific combination with a name and be able to select the variation among all the available ones directly in the rules management screen.

What do you think?

**Comment by otykier** [comment]
Would it make sense to add the required logic to the rules themselves?

For example, if you want a rule to only apply to Analysis Services models, you could add `Database.CompatibilityMode = "AnalysisServices"` to the Dynamic LINQ expression of the rule.

**Reply by nexus150** [reply]
It seems to me that it makes sense and is in line with the minimum compatibility level , which is already there,  it would be important that there be an option to include or exclude and that Power BI has 2 options, Pro (WithOut XMLA Endpoint) and Premium (With XMLA Endpoint)

**Reply by otykier** [reply]
Determining if a rule should be included or excluded bases on the CompatibilityMode is straightforward. Unfortunately, there is no way (to my knowledge) to detect from the model metadata, whether a model is in Pro (no XMLA) or Premium (with XMLA) capacity, so I don't think it will be possible to set up rules that can distinguish between these two scenarios.

**Comment by nexus150** [comment]
And what if the user selects which is the platform?  more or less similar to the original idea, but instead of being perspectives style it is much simpler.

**Reply by otykier** [reply]
In general, if the user needs to add information to a model, which is not part of the [Tabular Object Model](https://docs.microsoft.com/en-us/analysis-services/tom/introduction-to-the-tabular-object-model-tom-in-analysis-services-amo?view=asallproducts-allversions), there are two options: Annotations or Extended Properties. Both can be accessed by BPA rules.

So for example, if you create an annotation on the model with the name "PBIPlatform" and set its value to "Pro" or "Premium", you could use the following code in a BPA rule, to detect if the annotation is set to a specific value:

```
Model.GetAnnotation("PBIPlatform") = "Premium"
```
